Docker unit test TestChtimes fails on 32bit ARM server

Description of problem:
When running the Docker unit tests on an ARM 32bit server I get the following error

```
19:54:43 + go test -test.timeout=120m github.com/docker/docker/pkg/system
19:54:44 --- FAIL: TestChtimes (0.00s)
19:54:44 	chtimes_test.go:86: Expected: 2242-03-16 12:56:31 +0000 UTC, got: 1969-12-31 23:59:59 +0000 UTC
19:54:44 --- FAIL: TestChtimesLinux (0.00s)
19:54:44 	chtimes_unix_test.go:77: Expected: 2242-03-16 12:56:31 +0000 UTC, got: 1969-12-31 23:59:59 +0000 UTC
19:54:44 FAIL
19:54:44 coverage: 45.3% of statements
19:54:44 
19:54:44 Tests failed: ./pkg/system
```
It seems like these tests were introduced with the following commit https://github.com/docker/docker/commit/40b77af234319f02029368732249c2de0babb380

Testing this from the current docker/docker master branch with 1.9.0-dev.

@darrenstahlmsft Why exactly are you using 33 in "unixEpochTime.Add((1<<33 - 1) * time.Second)"?

`docker version`:
Client version: 1.6.2
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 7c8fca2-dirty
OS/Arch (client): linux/arm
Server version: 1.6.2
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 7c8fca2-dirty
OS/Arch (server): linux/arm


`docker info`:
Containers: 1
Images: 46
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 48
 Dirperm1 Supported: true
Execution Driver: native-0.2
Kernel Version: 4.2.5-260
Operating System: Ubuntu 15.04
CPUs: 4
Total Memory: 1.974 GiB
Name: scw-01
ID: 6PUR:TCAR:3PF3:6BKA:DWFL:PQS6:2JMH:XUTO:BLZF:JTRK:M4QK:LPFZ

`uname -a`:
Linux scw-01 4.2.5-260 #1 SMP Tue Oct 27 13:32:30 UTC 2015 armv7l armv7l armv7l GNU/Linux

Environment details (AWS, VirtualBox, physical, etc.):
Scaleway ARM Server, physical

Steps to Reproduce:
1. make test-unit
2.
3.

Actual Results:
1969-12-31 23:59:59 +0000 UTC

Expected Results:
2242-03-16 12:56:31 +0000 UTC

Hey Darren, I did test your PR and got the following result:

```
15:30:55 + go test -test.timeout=120m github.com/docker/docker/pkg/system
15:30:55 --- FAIL: TestChtimes (0.00s)
15:30:55 	chtimes_test.go:88: Expected: 2262-04-11 23:47:00 +0000 UTC, got: 1990-01-27 10:50:00 +0000 UTC
15:30:55 --- FAIL: TestChtimesLinux (0.00s)
15:30:55 	chtimes_unix_test.go:77: Expected: 2262-04-11 23:47:00 +0000 UTC, got: 1990-01-27 10:50:00 +0000 UTC
15:30:55 FAIL
15:30:55 coverage: 45.3% of statements
15:30:55 
15:30:55 Tests failed: ./pkg/system
```

--- a/pkg/system/chtimes.go
+++ b/pkg/system/chtimes.go
@@ -2,14 +2,30 @@ package system
 
 import (
 	"os"
+	"syscall"
 	"time"
+	"unsafe"
 )
+
+var (
+	maxTime time.Time
+)
+
+func init() {
+	if unsafe.Sizeof(syscall.Timespec{}.Nsec) == 8 {
+		// This is a 64 bit timespec
+		// os.Chtimes limits time to the following
+		maxTime = time.Unix(0, 1<<63-1)
+	} else {
+		// This is a 32 bit timespec
+		maxTime = time.Unix(1<<31-1, 0)
+	}
+}
 
 // Chtimes changes the access time and modified time of a file at the given path
 func Chtimes(name string, atime time.Time, mtime time.Time) error {
 	unixMinTime := time.Unix(0, 0)
-	// The max Unix time is 33 bits set
-	unixMaxTime := unixMinTime.Add((1<<33 - 1) * time.Second)
+	unixMaxTime := maxTime
 
 	// If the modified time is prior to the Unix Epoch, or after the
 	// end of Unix Time, os.Chtimes has undefined behavior

--- a/pkg/system/chtimes_test.go
+++ b/pkg/system/chtimes_test.go
@@ -30,9 +30,7 @@ func TestChtimes(t *testing.T) {
 	beforeUnixEpochTime := time.Unix(0, 0).Add(-100 * time.Second)
 	unixEpochTime := time.Unix(0, 0)
 	afterUnixEpochTime := time.Unix(100, 0)
-	// The max Unix time is 33 bits set
-	unixMaxTime := unixEpochTime.Add((1<<33 - 1) * time.Second)
-	afterUnixMaxTime := unixMaxTime.Add(100 * time.Second)
+	unixMaxTime := maxTime
 
 	// Test both aTime and mTime set to Unix Epoch
 	Chtimes(file, unixEpochTime, unixEpochTime)
@@ -90,31 +88,7 @@ func TestChtimes(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	if f.ModTime() != unixMaxTime {
-		t.Fatalf("Expected: %s, got: %s", unixMaxTime, f.ModTime())
-	}
-
-	// Test aTime after Unix max time and mTime set to Unix max time
-	Chtimes(file, afterUnixMaxTime, unixMaxTime)
-
-	f, err = os.Stat(file)
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	if f.ModTime() != unixMaxTime {
-		t.Fatalf("Expected: %s, got: %s", unixMaxTime, f.ModTime())
-	}
-
-	// Test aTime set to Unix Epoch and mTime before Unix Epoch
-	Chtimes(file, unixMaxTime, afterUnixMaxTime)
-
-	f, err = os.Stat(file)
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	if f.ModTime() != unixEpochTime {
-		t.Fatalf("Expected: %s, got: %s", unixEpochTime, f.ModTime())
+	if f.ModTime().Truncate(time.Second) != unixMaxTime.Truncate(time.Second) {
+		t.Fatalf("Expected: %s, got: %s", unixMaxTime.Truncate(time.Second), f.ModTime().Truncate(time.Second))
 	}
 }

--- a/pkg/system/chtimes_unix_test.go
+++ b/pkg/system/chtimes_unix_test.go
@@ -17,9 +17,7 @@ func TestChtimesLinux(t *testing.T) {
 	beforeUnixEpochTime := time.Unix(0, 0).Add(-100 * time.Second)
 	unixEpochTime := time.Unix(0, 0)
 	afterUnixEpochTime := time.Unix(100, 0)
-	// The max Unix time is 33 bits set
-	unixMaxTime := unixEpochTime.Add((1<<33 - 1) * time.Second)
-	afterUnixMaxTime := unixMaxTime.Add(100 * time.Second)
+	unixMaxTime := maxTime
 
 	// Test both aTime and mTime set to Unix Epoch
 	Chtimes(file, unixEpochTime, unixEpochTime)
@@ -87,35 +85,7 @@ func TestChtimesLinux(t *testing.T) {
 
 	stat = f.Sys().(*syscall.Stat_t)
 	aTime = time.Unix(int64(stat.Atim.Sec), int64(stat.Atim.Nsec))
-	if aTime != unixMaxTime {
-		t.Fatalf("Expected: %s, got: %s", unixMaxTime, aTime)
-	}
-
-	// Test aTime after Unix max time and mTime set to Unix max time
-	Chtimes(file, afterUnixMaxTime, unixMaxTime)
-
-	f, err = os.Stat(file)
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	stat = f.Sys().(*syscall.Stat_t)
-	aTime = time.Unix(int64(stat.Atim.Sec), int64(stat.Atim.Nsec))
-	if aTime != unixEpochTime {
-		t.Fatalf("Expected: %s, got: %s", unixEpochTime, aTime)
-	}
-
-	// Test aTime set to Unix Epoch and mTime before Unix Epoch
-	Chtimes(file, unixMaxTime, afterUnixMaxTime)
-
-	f, err = os.Stat(file)
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	stat = f.Sys().(*syscall.Stat_t)
-	aTime = time.Unix(int64(stat.Atim.Sec), int64(stat.Atim.Nsec))
-	if aTime != unixMaxTime {
-		t.Fatalf("Expected: %s, got: %s", unixMaxTime, aTime)
+	if aTime.Truncate(time.Second) != unixMaxTime.Truncate(time.Second) {
+		t.Fatalf("Expected: %s, got: %s", unixMaxTime.Truncate(time.Second), aTime.Truncate(time.Second))
 	}
 }

--- a/pkg/system/chtimes_windows_test.go
+++ b/pkg/system/chtimes_windows_test.go
@@ -17,9 +17,7 @@ func TestChtimesWindows(t *testing.T) {
 	beforeUnixEpochTime := time.Unix(0, 0).Add(-100 * time.Second)
 	unixEpochTime := time.Unix(0, 0)
 	afterUnixEpochTime := time.Unix(100, 0)
-	// The max Unix time is 33 bits set
-	unixMaxTime := unixEpochTime.Add((1<<33 - 1) * time.Second)
-	afterUnixMaxTime := unixMaxTime.Add(100 * time.Second)
+	unixMaxTime := maxTime
 
 	// Test both aTime and mTime set to Unix Epoch
 	Chtimes(file, unixEpochTime, unixEpochTime)
@@ -82,33 +80,7 @@ func TestChtimesWindows(t *testing.T) {
 	}
 
 	aTime = time.Unix(0, f.Sys().(*syscall.Win32FileAttributeData).LastAccessTime.Nanoseconds())
-	if aTime != unixMaxTime {
-		t.Fatalf("Expected: %s, got: %s", unixMaxTime, aTime)
-	}
-
-	// Test aTime after Unix max time and mTime set to Unix max time
-	Chtimes(file, afterUnixMaxTime, unixMaxTime)
-
-	f, err = os.Stat(file)
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	aTime = time.Unix(0, f.Sys().(*syscall.Win32FileAttributeData).LastAccessTime.Nanoseconds())
-	if aTime != unixEpochTime {
-		t.Fatalf("Expected: %s, got: %s", unixEpochTime, aTime)
-	}
-
-	// Test aTime set to Unix Epoch and mTime before Unix Epoch
-	Chtimes(file, unixMaxTime, afterUnixMaxTime)
-
-	f, err = os.Stat(file)
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	aTime = time.Unix(0, f.Sys().(*syscall.Win32FileAttributeData).LastAccessTime.Nanoseconds())
-	if aTime != unixMaxTime {
-		t.Fatalf("Expected: %s, got: %s", unixMaxTime, aTime)
+	if aTime.Truncate(time.Second) != unixMaxTime.Truncate(time.Second) {
+		t.Fatalf("Expected: %s, got: %s", unixMaxTime.Truncate(time.Second), aTime.Truncate(time.Second))
 	}
 }

